storybookjs/storybook
# Addon-controls: Improve color picker UI and migrate to react-colorful

Issue: N/A

## What I did

Migrates `addon-knobs` and `addon-controls` from [`react-color`](https://casesandberg.github.io/react-color/) to [`react-colorful`](https://github.com/omgovich/react-colorful).

Also added a "color" control to `official-storybook` examples so that it's easier to review and test this change.

The UI is slightly different

**Before:**

<img width="1199" alt="Screen Shot 2021-03-16 at 11 36 56 AM" src="https://user-images.githubusercontent.com/226640/111364019-f969db00-864d-11eb-9617-9125989097df.png">

**After:**

![Screen Shot 2021-03-16 at 11 03 59 AM](https://user-images.githubusercontent.com/226640/111364031-fc64cb80-864d-11eb-8a6e-abc2482cf361.png)

### Why?

`react-color` is a giant library (143kb) largely caused by the fact that it depends on `@icons/material` which brings in an SVG file for every supported icon in Google's Material Design. This is overkill for what Storybook needs, and the `react-colorful` library is designed to deliberately be lightweight and fast.

For us at [ZenHub](https://www.zenhub.com), we've noticed that the `react-color` attributed **greatly** to the initial installation performance of Storybook and it impacts every single one of our CI builds. In fact, `react-color` adds so many extra files to `node_modules` that our CI runners often run into file handle limit restrictions when installing, building and deploying Storybook.

Given that Storybook 6.2 has an emphasis on performance, I thought this change would be a nice small win. It's a small trade-off between UI elegance/flexibility and performance (specifically bundle size).

## Impact

Size of `node_modules` before this change: 140,107,104 bytes (236 MB on disk) for 34,882 items
Size of `node_modules` after this change: 132,129,715 bytes (212.3 MB on disk) for 29,389 items

26% fewer files used. 10% smaller install size.

It's a bit harder to measure, but I imagine the start up time for Storybook's webpack server should be faster too as Webpack has much fewer files to load into memory.

## How to test

- Run the `official-storybook` preview (`cd examples/official-storybook && yarn storybook`)
- Navigate to http://localhost:9011/?path=/story/addons-controls--basic and confirm that the Color control in the Controls tab works
- Navigate to http://localhost:9011/?path=/story/addons-knobs-withknobs--tweaks-static-values-organized-in-groups, go into the Knobs tab, go into the Display tab and ensure the Color control works
    - **NOTE:** The current `next` builds seem to have broken knobs. They don't have values. So you can't test that the change works, but you can test that the component renders properly
